src/Gaufrette/Adapter/Ftp.php

```diff
@@ -54,93 +55,122 @@ public function __construct($directory, $host, $options = array())
-
-        $path = $this->computePath($key);
-        $directory = dirname($path);
-
-        $this->ensureDirectoryExists($directory, true);
+        //change to the directory that will contain the written file
+        $this->moveToTargetDirectory($key);
+        $file = basename($key);
 
         $temp = fopen('php://temp', 'r+');
         $size = fwrite($temp, $content);
         rewind($temp);
 
-        if (!ftp_fput($this->getConnection(), $path, $temp, $this->mode)) {
+        if (!ftp_fput($this->getConnection(), $file, $temp, $this->mode)) {
+            //change back to ftp root directory
+            $this->changeDirectory($this->directory);
             fclose($temp);
 
             return false;
         }
 
+        //change back to ftp root directory
+        $this->changeDirectory($this->directory);
         fclose($temp);
 
         return $size;
     }
 
     /**
-     * {@inheritDoc}
+     * @param string $sourceKey
+     * @param string $targetKey
+     * @return bool
      */
     public function rename($sourceKey, $targetKey)
     {
-        $this->ensureDirectoryExists($this->directory, $this->create);
-
         $sourcePath = $this->computePath($sourceKey);
         $targetPath = $this->computePath($targetKey);
+        $targetDirectory = dirname($targetPath);
+
+        $this->ensureDirectoryExists($targetDirectory, true);
 
-        $this->ensureDirectoryExists(dirname($targetPath), true);
+        if (dirname($sourcePath) == $targetDirectory) {
+            //change to the directory that contains the file to rename
+            $this->changeDirectory($targetDirectory);
 
-        return ftp_rename($this->getConnection(), $sourcePath, $targetPath);
+            $result = ftp_rename($this->getConnection(), basename($sourceKey), basename($targetKey));
+
+            //change back to ftp root directory
+            $this->changeDirectory($this->directory);
+        } else {
+            // don't change current ftp directory. May fail if paths are too long
+            $result = ftp_rename($this->getConnection(), $sourcePath, $targetPath);
+        }
+
+        return $result;
     }
 
     /**
-     * {@inheritDoc}
+     * @param string $key
+     * @return bool
      */
     public function exists($key)
     {
-        $this->ensureDirectoryExists($this->directory, $this->create);
+        //change to the directory that contain the source file
+        $this->moveToTargetDirectory($key);
 
-        $file  = $this->computePath($key);
-        $lines = ftp_rawlist($this->getConnection(), '-al ' . dirname($file));
+        $lines = ftp_rawlist($this->getConnection(), '-al ' . '.');
 
         if (false === $lines) {
             return false;
         }
 
-        $pattern = '{(?<!->) '.preg_quote(basename($file)).'( -> |$)}m';
+        $pattern = '{(?<!->) '.preg_quote(basename($key)).'( -> |$)}m';
         foreach ($lines as $line) {
             if (preg_match($pattern, $line)) {
```

Comment:
this matching is broken now, as ``$line`` will not contain the full path anymore